libinteract/libinteract.py

```diff
@@ -866,6 +866,157 @@ def calc_cg_fullmatrix(identifiers, idxs, percmat, perco):
+            if filtered_array is not None:
+                array_list.append(filtered_array)
+        # Create vector of all nodes that are in contact with different chains
+        logical_vector = array_T[0] != array_T[sec_res]
+        if logical_vector.sum == 0:
+            warnstr = "No interchain contacts found" 
+            log.warning(warnstr)
+        else:
+            # For all combinations of different chains, find nodes that are in
+            # contact with different chains
+            for chain1, chain2 in itertools.combinations(chains, 2):
+                # Check both combinations (e.g. A, B and B, A)
+                filtered_array1 = filter_by_chain(chain1, chain2, sec_res, \
+                                               array, array_T)
+                filtered_array2 = filter_by_chain(chain2, chain1, sec_res, \
+                                               array, array_T)
+                # If both outputs exist, concatenate and append
+                if filtered_array1 is not None and filtered_array2 is not None:
+                    filtered_array = np.concatenate((filtered_array1, \
+                                                     filtered_array2))
+                    array_list.append(filtered_array)
+                # Only add (A, B)
+                elif filtered_array1 is not None:
+                    array_list.append(filtered_array1)
+                # Only add (B, A)
+                else:
+                    array_list.append(filtered_array2)
+    return array_list
+
+def create_matrix_list(fullmatrix, table_list, pdb, hb = False):
+    """Takes in the full matrix and returns a list of matrices. The
+    first matrix is the full matrix while the following matrices contain
+    intrachain persistences (if present) and interchain persistences 
+    (if present).
+    """
+
+    # Initialize output
+    mat_list = [fullmatrix]
+    # Select which column has the resid
+    if hb:
+        sec_res_id = 5
+    else:
+        sec_res_id = 4
+    # If contacts exist in multiple chains
+    if len(table_list) > 1:
+        # Map each residue id to a position in the matrix
+        resids = pdb.residues.resids
+        res_dict = {str(resids[i]):i for i in range(fullmatrix.shape[0])}
+        # For all tables excluding the first one
+        for i in range(1, len(table_list)):
+            # Create empty matrix
+            matrix = np.zeros(fullmatrix.shape)
+            # Get list of indexes for the matrix
+            table = table_list[i].T
+            mat_i = [res_dict[i] for i in table[1]]
+            mat_j = [res_dict[j] for j in table[sec_res_id]]
+            # Fill the matrix with appropriate values
+            matrix[mat_i, mat_j] = fullmatrix[mat_i, mat_j]
+            matrix[mat_j, mat_i] = fullmatrix[mat_j, mat_i]
+            mat_list.append(matrix)
+    return mat_list
+
+
+def save_output_list(table_list, filename, mat_list=None, hb=False):
```

Comment:
I find the function definition here slightly confusing - I don't like the fact that it returns different things depending on whether the `mat_list` argument, which is optional, is set or not. If you implement the changes I suggested above (i.e. using a dictionary instead of the table/matrix list) you can have both `table_list` and `mat_list` as optional mutually exclusive arguments (and so that the function raises an error if you supply both) which is a bit more streamlined